content/concepts/consensus-network/invariant-checking.md

```diff
@@ -0,0 +1,404 @@
+# Invariant Checking
+
+This article provides a high-level overview of Invariant Checking, why it exist, how it works, and lists active invariant checks.
+
+When building applications on the XRP Ledger, it is crucial to understand the idea of Invariant Checks so as not to be surprised by specific error codes.
+
+## Introduction
+
+Invariant checking is a protection enhancement to the XRP Ledger and reinforces the critical properties of the XRP Ledger.
+
+Invariant checks should not trigger, but they ensure the XRP Ledger's integrity from bugs yet to be discovered or even created.
+
+
+## Why it Exists
+
+Nevertheless, why do we need invariant checks?
+
+- The overarching code for the XRP Ledger is complicated and vast; therefore, there is a high potential for code to execute incorrectly.
+- The cost of incorrectly executing a transaction is high and not acceptable by any standards.
+
+
+## How it Works
+
+A second layer of code runs automatically and in real-time after each transaction completes. It then examines the changes it made for correctness before the results are committed to the ledger. After every transaction runs, the invariant checker runs. Problematic transactions are marked with a **tecINVARIANT_FAILED** result code and are included in the ledger as having done nothing.
+
+The invariant checker is then invariantly checked to ensure the invariant checker's failure is not due to a bug. If this second level check fails, transactions are marked with a **tefINVARIANT_FAILED** result code. The transaction is in its final state and is not included in the ledger.
+
+
+## Active Invariant Checks
+
+Following is a list of active checks that the invariant checker runs against each transaction on the XRP Ledger.
+
+- [Transaction Fee Check](#transaction-fee-check)
+- [XRP Not Created](#xrp-not-created)
+- [Account Roots Not Deleted](#account-roots-not-deleted)
+- [XRP Balance Checks](#xrp-balance-checks)
+- [Ledger Entry Types Match](#ledger-entry-types-match)
+- [No XRP Trust Lines](#no-xrp-trust-lines)
+- [No Bad Offers](#no-bad-offers)
+- [No Zero Escrow](#no-zero-escrow)
+- [Valid New Account Root](#valid-new-account-root)
+
+
+### Transaction Fee Check
+
+- **Condition(s) Checked:**
+
+    - [Trnsaction fees](https://xrpl.org/rippleapi-reference.html#transaction-fees) should never be negative nor larger than the transaction itself.
+
+```
+class TransactionFeeCheck
+{
+public:
+    void
+    visitEntry(
+        bool,
+        std::shared_ptr<SLE const> const&,
+        std::shared_ptr<SLE const> const&);
+
+    bool
+    finalize(
+        STTx const&,
+        TER const,
+        XRPAmount const,
+        ReadView const&,
+        beast::Journal const&);
+};
+```
+
+### XRP Not Created
+
+- **Condition(s) Checked:**
+    
+    - A transaction must not create XRP and should only destroy the XRP [fee](https://xrpl.org/rippleapi-reference.html#transaction-fees)).
```

Comment:
1. As mentioned above, please use a relative URL, not an absolute/fully-qualified URL here.
2. In the documentation we try to avoid using the word "fee" for the [transaction cost](https://xrpl.org/transaction-cost.html) because people tend to assume that a fee is paid to someone in particular (it's actually destroyed). Also, that's the more relevant concept link for this idea.


```suggestion
    - A transaction must not create XRP and should only destroy the XRP [transaction cost](transaction-cost.html)).
```